services/requester/local_client.go

```diff
@@ -0,0 +1,100 @@
+package requester
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/onflow/go-ethereum/common"
+
+	"github.com/onflow/flow-evm-gateway/services/state"
+	"github.com/onflow/flow-evm-gateway/storage"
+)
+
+var _ EVMClient = &LocalClient{}
+
+func NewLocalClient(state *state.BlockState, blocks storage.BlockIndexer) *LocalClient {
+	return &LocalClient{
+		state:  state,
+		blocks: blocks,
+	}
+}
+
+type LocalClient struct {
+	state  *state.BlockState
+	blocks storage.BlockIndexer
+}
+
+func (l *LocalClient) SendRawTransaction(
+	ctx context.Context,
+	data []byte,
+) (common.Hash, error) {
+	return common.Hash{}, fmt.Errorf("local client is read-only")
+}
+
+func (l *LocalClient) GetBalance(
+	ctx context.Context,
+	address common.Address,
+	evmHeight int64,
+) (*big.Int, error) {
+	bal := l.state.GetBalance(address)
+	return (&big.Int{}).SetUint64(bal.Uint64()), nil
+}
+
+func (l *LocalClient) Call(
+	ctx context.Context,
+	data []byte,
+	from common.Address,
+	evmHeight int64,
+) ([]byte, error) {
+	res, err := l.state.Call(from, data)
+	if err != nil {
+		return nil, err
+	}
+
+	// todo what if it failed?
+
+	return res.ReturnedData, nil
+}
+
+func (l *LocalClient) EstimateGas(
+	ctx context.Context,
+	data []byte,
+	from common.Address,
+	evmHeight int64,
+) (uint64, error) {
+	res, err := l.state.Call(from, data)
```

Comment:
I am not sure that we can rely on the `res.GasConsumed` for the gas estimation, because the `EVM.dryRun` that we currently use, adds some buffer to cover for various cases: https://github.com/onflow/flow-go/blob/96539065a18bc36a88475b4fafd6c7d276b6f6ed/fvm/evm/emulator/emulator.go#L283-L304. I guess we may have to use the gas estimator logic that Geth also uses, but not sure it should be part of this PR.